### ISSUE TITLE
better cloning for the binding formvalue -> model

### ISSUE THREAD

**Issue by lucienbertin** [ngx-formly/ngx-formly#1719]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request
```

**Current behavior**
since version 5.0, when binding the form.value to the model, you now `clone` the current form value to assign it to the model

https://github.com/ngx-formly/ngx-formly/blob/master/src/core/src/lib/components/formly.form.ts#L83
```ts
  changeModel(event: { key: string, value: any }) {
    assignModelValue(this.model, event.key.split('.'), event.value);
    this.modelChange.emit(clone(this.model));
  }
```
while previous versions would not go through a clone
v4 - https://github.com/ngx-formly/ngx-formly/blob/v4/src/core/src/lib/components/formly.form.ts#L71
```ts
  changeModel(event: { key: string, value: any }) {
    assignModelValue(this.model, event.key, event.value);
    this.modelChange.emit(this.model);
    this.checkExpressionChange();
  }
```

_Now_, i have no beef with going through a clone, but my problem is that when you clone objects, you just copy properties and getters, and not all the methods. as a result, when i use a moment for one of my properties, then after the cloning the object is not a moment anymore and i cant use `.format()` or other methods.

a solution would be to patch the method clone, so that objects that know how to clone themselves do it, instead of you copying each property

https://github.com/ngx-formly/ngx-formly/blob/master/src/core/src/lib/utils.ts#L104
```ts
export function clone(value: any): any {
  if (!isObject(value) || value instanceof RegExp || value instanceof Observable || /* instanceof SafeHtmlImpl */ value.changingThisBreaksApplicationSecurity) {
    return value;
  }

  if (value instanceof AbstractControl) {
    return null;
  }

  if (value instanceof Date) {
    return new Date(value.getTime());
  }

  if (Array.isArray(value)) {
    return value.slice(0).map(v => clone(v));
  }

  //add this
  if (value.clone && value.clone()) {
    return value.clone();
  }

  return Object.keys(value).reduce((newVal, prop) => {
    const propDescriptor = Object.getOwnPropertyDescriptor(value, prop);

    if (propDescriptor.get) {
      Object.defineProperty(newVal, prop, { ...propDescriptor, get: () => clone(value[prop]) });
    } else {
      newVal[prop] = clone(value[prop]);
    }

    return newVal;
  }, {});
}
```
(fork and PR incoming)
**Expected behavior**
when cloning a property that is a `moment`, it stays a moment.

**Minimal reproduction of the problem with instructions**
https://stackblitz.com/edit/angular-to6hpw

when you set a date and submit, it'll try to call `.format()` on the form value and model value

[app.component.ts#L32](https://stackblitz.com/edit/angular-to6hpw?file=src%2Fapp%2Fapp.component.ts)
```ts
  submit() {
    console.log(`form.value.date.format(): ${this.form.value.date.format()}`);
    console.log(`model.date.format(): ${this.model.date.format()}`); // throws
  }
```

the first console.log will operate normally, but the second will throw because `model.date` is not a moment anymore

**What is the motivation / use case for changing the behavior?**
one of the great feature of formly is that i can define complexe input types for complex objects, and then handle them like any other field in a form. it's a great for respecting the [SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle) as the edition of the object is the sole responsibility of the formly field i coded.

with this cloning that loses the class of the different properties, i cant be sure that my properties have the right type

**Please tell us about your environment:**
windows 10
vscode
npm

* **Angular version:** 8.x

* **Browser:** all

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]

* **Node (for AoT issues):** `node --version` = lts - 10.16.2 at time of writing

**Comment by swillisstudio** [comment]
Been pulling my hair out with this exact same issue, glad to see I'm not the only one! Excited for this pull request to be merged in.

**Comment by aitboudad** [comment]
released in `5.4` version.
